[#7177] feat(core): Add a Caffeine-based implementation for EntityCache

Abyss-lord opened the pull request:
### What changes were proposed in this pull request?

Add a Caffeine-based implementation for EntityCache.

- [X] Implement the methods related to the `SupportsEntityStoreCache` and `SupportsRelationEntityCache` interface in `CaffeineEntityCache`.
- [ ] Add unit tests for both index and cache.
- [ ] Use JCStress to perform multi-threaded testing on `CaffeineEntityCache` and the `CacheIndex`.

### Why are the changes needed?

Fix: #7177

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

local test.

xunliu commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -0,0 +1,179 @@
+  public static <E extends Entity & HasIdentifier> Class<E> getEntityClass(Entity.EntityType type) {
+    Preconditions.checkNotNull(type, "EntityType must not be null");
+
+    Class<?> aClass = ENTITY_CLASS_MAP.get(type);
```
Maybe the `clazz` is better than `aClass`?

Abyss-lord replied:
fix

xunliu commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -0,0 +1,179 @@
+   * @return The {@link NameIdentifier} of the metadata
+   */
+  protected static NameIdentifier getIdentFromMetadata(Entity metadata) {
+
```
Please remove this blank line.

Abyss-lord replied:
Fix

xunliu commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -0,0 +1,179 @@
+    return (List<E>) (List<? extends Entity>) entities;
+  }
+
+  /**
+   * Converts an entity to a new entity.
+   *
+   * @param entity The original entity.
+   * @return A new entity.
+   * @param <E> The type of the entity.
+   */
+  @SuppressWarnings("unchecked")
+  protected static <E extends Entity & HasIdentifier> E convertSafe(Entity entity) {
+    if (!(entity instanceof HasIdentifier)) {
+      throw new IllegalStateException(
+          "Cached entity " + entity + " is not of expected type: " + HasIdentifier.class.getName());
+    }
+
+    return (E) entity;
+  }
```
I think maybe the `convertEntity` is better than `converSafe`

Abyss-lord replied:
FIx

xunliu commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -0,0 +1,179 @@
+   * @return A list of entities.
+   * @param <E> The type of the elements in the list.
+   */
+  protected static <E extends Entity & HasIdentifier> List<Entity> toEntityList(
+      List<E> sourceList) {
+    return sourceList.stream().map(e -> (Entity) e).collect(Collectors.toList());
+  }
```
I find the `toEntityList()` only calls once, I think we didn't need to create a function, and this function is very simple

Abyss-lord replied:
Fix

xunliu commented on core/src/main/java/org/apache/gravitino/cache/CacheConfig.java:
```diff
@@ -0,0 +1,187 @@
+
+package org.apache.gravitino.cache;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
```
Please remove this duplicate license header.

Abyss-lord replied:
Fix

xunliu commented on core/src/main/java/org/apache/gravitino/cache/CacheConfig.java:
```diff
@@ -0,0 +1,187 @@
+ * expiration settings.
+ */
+public class CacheConfig extends Config {
+
```
Please remote this blank line

Abyss-lord replied:
Fix

xunliu commented on core/src/main/java/org/apache/gravitino/cache/CacheConfig.java:
```diff
@@ -0,0 +1,187 @@
+
+  // Maximum number of entries in the cache
+  public static final ConfigEntry<Integer> CACHE_MAX_SIZE =
+      new ConfigBuilder("gravitino.server.cache.max.size")
```
Maybe the `gravitino.server.cache.max.num` is better than `gravitino.server.cache.max.size`, because `size` lets us think of memory size?

Abyss-lord replied:
Fix

xunliu commented on core/src/main/java/org/apache/gravitino/cache/CacheConfig.java:
```diff
@@ -0,0 +1,187 @@
+          .booleanConf()
+          .createWithDefault(true);
+
+  // Maximum weight of cache entries
+  public static final ConfigEntry<Long> CACHE_MAX_WEIGHT =
+      new ConfigBuilder("gravitino.server.cache.max.weight")
+          .doc("The maximum weight of cache entries. default is 10000.")
+          .version(ConfigConstants.VERSION_0_10_0)
+          .longConf()
+          .checkValue(value -> value > 0, ConfigConstants.POSITIVE_NUMBER_ERROR_MSG)
+          .createWithDefault(EntityCacheWeigher.getMaxWeight());
```
I'm not sure we need this configuration.

Abyss-lord replied:
I already remove this configuration.

xunliu commented on core/src/main/java/org/apache/gravitino/cache/CaffeineEntityCache.java:
```diff
@@ -0,0 +1,458 @@
+    return INSTANCE;
+  }
+
+  /**
+   * Returns the instance of MetaCacheCaffeine, if it is initialized, otherwise throws an exception.
+   *
+   * @return If INSTANCE initialized, returns the instance, otherwise throws an exception.
+   */
+  public static CaffeineEntityCache getInstance() {
+    if (INSTANCE == null) {
+      throw new RuntimeException("Illegal state: instance not initialized");
+    }
+
+    return INSTANCE;
+  }
```
I think we didn't need this function.

Abyss-lord replied:
Fix, remove this function.

xunliu commented on core/src/main/java/org/apache/gravitino/cache/CaffeineEntityCache.java:
```diff
@@ -0,0 +1,458 @@
+              try {
+                invalidateExpiredItem(key);
+              } catch (Throwable t) {
+                LOG.error("Async removal failed for {}", value, t);
```
This log message is too vague

Abyss-lord replied:
Fix

xunliu commented on core/src/main/java/org/apache/gravitino/cache/CaffeineEntityCache.java:
```diff
@@ -0,0 +1,458 @@
+            return convertSafe(entitiesFromCache.get(0));
+          }
+
+          E entityFromDb = entityStore.get(ident, type, getEntityClass(type));
```
`entityFromStore`

Abyss-lord replied:
Fix

xunliu commented on core/src/main/java/org/apache/gravitino/cache/EntityCacheWeigher.java:
```diff
@@ -0,0 +1,103 @@
+ * weight is calculated as follows:
+ *
+ * <ul>
+ *   <li>Metalake: 100 bytes
+ *   <li>Catalog: 75 bytes
+ *   <li>Schema: 50 bytes
+ *   <li>Other: 15 bytes
```
What does Metalake: 100` bytes`?
You mean is percentage?

Abyss-lord replied:
Fix

Abyss-lord commented:
Hi @xunliu , I've completed the code updates and would appreciate your review of the PR when you have a moment. Here's a summary of the commits:

xunliu commented:
@Abyss-lord This PR only has UT, can you add ITs in the next PR?

Abyss-lord commented:
> @Abyss-lord This PR only has UT, can you add ITs in the next PR?

Yes. Because this PR is too large, the IT test will be added in the next PR.

jerryshao commented:
The code is still too large to review, can we just split to implement part of the APIs in this PR, and implement some other PRs in another PR.

Abyss-lord commented:
> The code is still too large to review, can we just split to implement part of the APIs in this PR, and implement some other PRs in another PR.

yes, I will do it.

jerryshao commented:
If we want to make the ball roll fast, we'd better control the PR size, otherwise it will cause the reviewer lots of effort and move slowly.

yuqi1129 commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -0,0 +1,165 @@
+  @SuppressWarnings("unchecked")
+  protected static <E extends Entity & HasIdentifier> List<E> convertEntity(List<Entity> entities) {
+    for (Entity e : entities) {
+      if (!(e instanceof HasIdentifier)) {
+        throw new IllegalStateException(
+            "Cached entity " + e + " is not of expected type: " + HasIdentifier.class.getName());
+      }
```
Maybe you can reuse the code in L139.

Abyss-lord replied:
Fix

yuqi1129 commented on gradle/libs.versions.toml:
```diff
@@ -119,6 +119,8 @@ hudi = "0.15.0"
 aliyun-credentials = "0.3.12"
 openlineage = "1.29.0"
+concurrent-trees = "2.6.0"
+jcstress = "0.4"
```
What is this used for?

Abyss-lord replied:
> What is this used for?

@yuqi1129 
- concurrent-trees：Provides a high-performance concurrent radix tree (prefix tree) implementation, use as cacheIndex
- jcstress: A concurrency testing tool.

yuqi1129 replied:
I see

Abyss-lord commented:
Hi @yuqi1129 , I've completed the code updates and would appreciate your review of the PR when you have a moment. Here's a summary of the commits:
1. reuse the codes.

Abyss-lord commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -0,0 +1,168 @@
+   * @return The {@link NameIdentifier} of the metadata
+   */
+  protected static NameIdentifier getIdentFromMetadata(Entity metadata) {
+    if (metadata instanceof HasIdentifier) {
+      HasIdentifier hasIdentifier = (HasIdentifier) metadata;
+      return hasIdentifier.nameIdentifier();
+    }
+
+    throw new IllegalArgumentException("Unsupported EntityType: " + metadata.type().getShortName());
```
Fix

yuqi1129 commented on core/src/main/java/org/apache/gravitino/cache/EntityCacheWeigher.java:
```diff
@@ -0,0 +1,123 @@
+   * @return The maximum weight that can be stored in the cache.
+   */
+  public static long getMaxWeight() {
+    return 2
```
I mean we can introduce a static value  `MAX_WEIGHT = 2 * (xxxxxx);` and then in this value we can directly use this value.

Abyss-lord commented:
Hi @yuqi1129 , I've completed the code updates and would appreciate your review of the PR when you have a moment. Here's a summary of the commits:
1. Introduce a static value  MAX_WEIGHT.

jerryshao commented on core/build.gradle.kts:
```diff
@@ -37,6 +38,7 @@ dependencies {
   implementation(libs.guava)
   implementation(libs.h2db)
   implementation(libs.mybatis)
+  implementation(libs.concurrent.trees)
```
If there's other libraries you introduced, you need list all of them into the LICENSE.bin.

Abyss-lord replied:
- `concurrent-trees` is licensed under Apache License 2.0, which is compatible with our project license.
- `jcstress` is licensed under GPL-2.0.

@jerryshao I will update the LICENSE.bin to include both dependencies.

Abyss-lord replied:
FIX

Abyss-lord replied:
Hi @jerryshao , I've completed the code updates and would appreciate your review of the PR when you have a moment. Here's a summary of the commits:
1. Add LICENSE.bin.

justinmclean replied:
It depends on how it is used, if it is an optional dependency, i.e only used with testing, then that is OK.

Abyss-lord replied:
@justinmclean Thanks for the clarification! `concurrent-trees` is a runtime dependency and will be used in the core implementation.However, `jcstress` is only used for concurrency coherence testing and will not be included in the runtime binary. It's scoped strictly to the test framework to validate thread-safety and consistency under concurrent access.

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -0,0 +1,166 @@
+   */
+  @SuppressWarnings("unchecked")
+  public static <E extends Entity & HasIdentifier> Class<E> getEntityClass(Entity.EntityType type) {
+    Preconditions.checkNotNull(type, "EntityType must not be null");
```
Don't use `checkNotNull`, it will throw NPE, I think I mentioned in different PR before.

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -0,0 +1,166 @@
+   * @param metadata The entity
+   * @return The {@link NameIdentifier} of the metadata
+   */
+  protected static NameIdentifier getIdentFromMetadata(Entity metadata) {
```
Change the variable name to `entity`, `entity` is an entity, `metadata` is just a concept.

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -0,0 +1,166 @@
+   */
+  protected static void validateEntityHasIdentifier(Entity entity) {
+    if (!(entity instanceof HasIdentifier)) {
+      throw new IllegalArgumentException("Unsupported EntityType: " + entity.type().getShortName());
```
short name is meaningless, you'd use the type as a string directly.

jerryshao replied:
Also you can simplify with `Preconditions.xxx`.

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -0,0 +1,166 @@
+    return (E) entity;
+  }
+
+  /**
+   * Constructs a new {@link BaseEntityCache} instance. If the provided entityStore is null, it will
+   * use the entity store configured in the Gravitino environment.
+   *
+   * @param entityStore The entity store to be used by the cache, can be null.
+   */
+  public BaseEntityCache(CacheConfig cacheConfig, EntityStore entityStore) {
+    this.cacheConfig = cacheConfig;
+    this.entityStore = (RelationalEntityStore) entityStore;
+  }
```
Move the public constructor to the top of the method.

Abyss-lord replied:
Fix

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/CacheConfig.java:
```diff
@@ -0,0 +1,158 @@
+  public static final ConfigEntry<Boolean> CACHE_WEIGHER_ENABLED =
+      new ConfigBuilder("gravitino.server.cache.weigher.enabled")
+          .doc(
+              "Whether to enable weighted cache. if enabled, cache entries will be weighed based on their weight, not"
```
I guess this line exceeds 100 characters, can you check and break the line.

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/CacheConfig.java:
```diff
@@ -0,0 +1,158 @@
+public class CacheConfig extends Config {
+  // Maximum number of entries in the cache
+  public static final ConfigEntry<Integer> CACHE_MAX_SIZE =
+      new ConfigBuilder("gravitino.server.cache.max.num")
```
We don't need 'server' in the configure name, also we typically use 3 level name, like "gravitino.cache.maxEntries", besides, it is better to put all the configs to `Configs` class, no need to create a class specifically.

Abyss-lord replied:
Fix

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/CacheConfig.java:
```diff
@@ -0,0 +1,158 @@
+
+  // Cache entry expiration time
+  public static final ConfigEntry<Long> CACHE_EXPIRATION_TIME =
+      new ConfigBuilder("gravitino.server.cache.expirationTimeMin")
```
expireTimeInMs, better to use MS so you can easily trigger the test.

Abyss-lord replied:
Fix

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/CaffeineEntityCache.java:
```diff
@@ -0,0 +1,481 @@
+  ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
+
+  @VisibleForTesting
+  public static void resetForTest() {
```
If it is only for test, you can make it package public.

Abyss-lord replied:
Fix

jerryshao commented on core/src/main/java/org/apache/gravitino/storage/relational/CachedEntityStore.java:
```diff
@@ -0,0 +1,254 @@
+import org.apache.gravitino.tag.SupportsTagOperations;
+import org.apache.gravitino.utils.Executable;
+
+public class CachedEntityStore
```
I think this part can be moved to another PR to decrease the size of this PR. Also, it is unrelated to the cache implementation in this PR.

Abyss-lord replied:
Fix

jerryshao commented on gradle/libs.versions.toml:
```diff
@@ -282,6 +284,9 @@ aliyun-credentials-sdk = { group='com.aliyun', name='credentials-java', version.
 openlineage-java= { group = "io.openlineage", name = "openlineage-java", version.ref = "openlineage" }
 
+concurrent-trees = {group="com.googlecode.concurrent-trees", name="concurrent-trees", version.ref="concurrent-trees"}
+jcstress = {group="org.openjdk.jcstress", name="jcstress-core", version.ref="jcstress"}
```
Leave the whitespace before and after like above.

jerryshao commented on core/src/test/java/org/apache/gravitino/cache/TestCacheConfig.java:
```diff
@@ -0,0 +1,39 @@
+    Assertions.assertTrue(cacheConfig.isWeigherEnabled());
+    Assertions.assertEquals(200_302_000, cacheConfig.getMaxWeight());
+  }
+}
```
You don't have tests for caffeine cache, only add the trivial tests for configuration, the test coverage is not enough.

Abyss-lord commented:
Hi @jerryshao , I've completed the code updates and would appreciate your review of the PR when you have a moment. Here's a summary of the commits:
1. Fixed all reported code issues
2. Updated configuration docs
	a. Ensured character length limits are documented
	b. Moved all config keys into the Configs class for better organization
3. Refactored and improved locking logic

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/CaffeineEntityCache.java:
```diff
@@ -0,0 +1,493 @@
+        opLock.unlock();
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```
Shall we interrupt the current thread, if so then this entity cache will not work any more, right?

Abyss-lord commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -0,0 +1,166 @@
+  public static <E extends Entity & HasIdentifier> Class<E> getEntityClass(Entity.EntityType type) {
+    Preconditions.checkArgument(type != null, "EntityType must not be null");
+
+    Class<?> clazz = ENTITY_CLASS_MAP.get(type);
+    if (clazz == null) {
+      throw new IllegalArgumentException("Unsupported EntityType: " + type.getShortName());
+    }
```
Fix.

Abyss-lord commented on core/src/main/java/org/apache/gravitino/cache/CaffeineEntityCache.java:
```diff
@@ -0,0 +1,493 @@
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean contains(
+      NameIdentifier ident, Entity.EntityType type, SupportsRelationOperations.Type relType) {
+    return withLock(() -> cacheData.getIfPresent(EntityCacheKey.of(ident, type)) != null);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean contains(NameIdentifier ident, Entity.EntityType type) {
+    return withLock(() -> cacheData.getIfPresent(EntityCacheKey.of(ident, type)) != null);
+  }
```
@jerryshao yes, these two method only Yes, these two methods only check the cache and do not access the underlying store.

Abyss-lord commented on core/src/main/java/org/apache/gravitino/cache/CaffeineEntityCache.java:
```diff
@@ -0,0 +1,493 @@
+        () -> {
+          EntityCacheKey entityCacheKey = EntityCacheKey.of(nameIdentifier, identType, relType);
+
+          List<Entity> entitiesFromCache = cacheData.getIfPresent(entityCacheKey);
+          if (entitiesFromCache != null) {
+            List<E> convertedEntities = convertEntity(entitiesFromCache);
+            return Optional.of(convertedEntities);
+          }
+
+          return Optional.empty();
```
Fix

Abyss-lord commented on core/src/main/java/org/apache/gravitino/Configs.java:
```diff
@@ -343,4 +343,48 @@ private Configs() {}
+          .createWithDefault(10_000);
+
+  // Whether to enable cache expiration
+  public static final ConfigEntry<Boolean> CACHE_EXPIRATION_ENABLED =
+      new ConfigBuilder("gravitino.cache.expiration.enabled")
+          .doc("Whether to enable cache entry expiration based on time.")
+          .version(ConfigConstants.VERSION_0_10_0)
+          .booleanConf()
+          .createWithDefault(true);
```
Fix

Abyss-lord commented on core/src/main/java/org/apache/gravitino/Configs.java:
```diff
@@ -343,4 +343,48 @@ private Configs() {}
+
+  // Whether to enable cache status logging
+  public static final ConfigEntry<Boolean> CACHE_STATUS_ENABLED =
+      new ConfigBuilder("gravitino.cache.status.enabled")
```
@jerryshao Fix, I've updated the config key to use the 3-level format, e.g. `gravitino.cache.enableWeigher` and `gravitino.cache.enableStats`.

Abyss-lord commented on core/src/main/java/org/apache/gravitino/cache/CaffeineEntityCache.java:
```diff
@@ -0,0 +1,493 @@
+     */
+    ThreadPoolExecutor cleanupExec =
+        new ThreadPoolExecutor(
+            1,
+            1,
+            0L,
+            TimeUnit.MILLISECONDS,
+            new ArrayBlockingQueue<>(100),
```
@jerryshao I've extracted the cleanup executor construction into a dedicated method `buildCleanupExecutor` and made the core thread size, max thread size, and queue capacity configurable via the following keys:

- `gravitino.cache.cleanup.coreThreads`
- `gravitino.cache.cleanup.maxThreads`
- `gravitino.cache.cleanup.queueCapacity`

jerryshao replied:
You don't have to add these configurations, it is useless for users, just don't hard code a number directly, using a constant to define this number.

Abyss-lord commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -0,0 +1,166 @@
+   * @param <E> The type of the entities in the list.
+   */
+  @SuppressWarnings("unchecked")
+  protected static <E extends Entity & HasIdentifier> List<E> convertEntity(List<Entity> entities) {
```
@jerryshao Thanks for pointing that out! I just measured the line and it’s 98 characters long, so it's still within the 100-character limit. Let me know if you’d prefer it wrapped anyway for readability.

Abyss-lord commented:
Hi @jerryshao , I've completed the code updates and would appreciate your review of the PR when you have a moment. Here's a summary of the commits:
1. Refactored thread pool construction: Moved the cleanup executor creation into a dedicated `buildCleanupExecutor` method for improved readability and reuse.
2. Style improvements；
3. Refactored configuration keys and documentation: Renamed several config entries and refined their doc strings for clarity and consistency.
4. Fixed logic issues.
5. Comment corrections.

jerryshao commented on core/src/main/java/org/apache/gravitino/Configs.java:
```diff
@@ -343,4 +343,50 @@ private Configs() {}
+              "Time in milliseconds after which a cache entry expires. Default is 3,600,000 ms (1 hour).")
+          .version(ConfigConstants.VERSION_0_10_0)
+          .longConf()
+          .checkValue(value -> value > 0, ConfigConstants.POSITIVE_NUMBER_ERROR_MSG)
```
You should allow setting 0 as a configuration value.

Abyss-lord replied:
> You should allow setting 0 as a configuration value.

Yes, it has been fixed.

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/CaffeineEntityCache.java:
```diff
@@ -0,0 +1,505 @@
+   */
+  private CaffeineEntityCache(Config cacheConfig, EntityStore entityStore) {
+    super(cacheConfig, entityStore);
+    cacheIndex = new ConcurrentRadixTree<>(new DefaultCharArrayNodeFactory());
```
Please add `this.`. For the first time class variable is initialized, it is better to add `this.` to the variable.

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/CaffeineEntityCache.java:
```diff
@@ -0,0 +1,493 @@
+    return withLockAndThrow(
+        () -> {
+          EntityCacheKey entityCacheKey = EntityCacheKey.of(ident, type);
+          List<Entity> entitiesFromCache = cacheData.getIfPresent(entityCacheKey);
```
You'd better add some comments to explain it.

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/EntityCacheWeigher.java:
```diff
@@ -0,0 +1,131 @@
+        break;
+    }
+
+    return weight;
```
I think you can maintain an immutable static map for entity type and weight, so you can directly the weight from the map, no need to use switch case to check.

Abyss-lord replied:
Fix
